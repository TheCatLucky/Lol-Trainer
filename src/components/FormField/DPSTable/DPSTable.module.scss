.wrapper {
  margin-top: 25px;

  .table {
    border: 1px solid black;
    border-spacing: 0;
    border-collapse: collapse;
    background-color: #f8f8f8;

    thead {
      tr {
        border-bottom: 1px solid black;
        background-color: lightyellow;

        td {
          text-align: center;

          &:first-child {
            border-right: 1px solid black;
          }
        }
      }
    }

    tbody {
      tr {
        border-bottom: 1px solid black;

        td {
          height: 30px;
          width: 60px;
          text-align: center;
          border-right: 1px solid black;

          &:first-child {
            width: 110px;
            background-color: lightyellow;
          }
        }
      }
    }
  }
}
