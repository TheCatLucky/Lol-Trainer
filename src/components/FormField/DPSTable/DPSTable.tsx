import { FC, useEffect, useState } from 'react';

import { ChampionModel } from '../../../models';
import classes from './DPSTable.module.scss';

type Props = {
  champion: ChampionModel;
};

const DPSTable: FC<Props> = (props) => {
  const { champion } = props;
  const {
    attackDamage,
    attackSpeed,
    critDamage,
    critChance,
    armorFlatPenetration,
    armorPenetration,
  } = champion.stats;
  const baseArmorResistance = [0, 10, 20, 30, 40, 50, 60, 70, 80, 90, 100];
  const [damage, setDamage] = useState([0]);
  const [criticalDamage, setCriticalDamage] = useState([0]);
  const [dps, setDps] = useState([0]);

  useEffect(() => {
    const nonCritDmg: number[] = [];
    const critDMG: number[] = [];
    const dpsDMG: number[] = [];

    baseArmorResistance.forEach((armor) => {
      const baseDamage =
        attackDamage *
        (100 / (100 + Math.max(armor * (1 - armorPenetration) - armorFlatPenetration, 0)));

      const defaultAttack = baseDamage;
      const critAttack = baseDamage * critDamage;

      nonCritDmg.push(Math.floor(defaultAttack));
      critDMG.push(Math.floor(critAttack));
      dpsDMG.push(
        Math.floor((defaultAttack * (1 - critChance) + critAttack * critChance) * attackSpeed),
      );
    });

    setDamage(nonCritDmg);
    setCriticalDamage(critDMG);
    setDps(dpsDMG);
  }, [champion]);

  return (
    <div className={classes.wrapper}>
      <table className={classes.table}>
        <thead>
          <tr>
            <td />
            <td colSpan={11}>Расчет урона в зависимости от сопортивлений</td>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>Армор</td>
            {
              baseArmorResistance.map((arm, index) => (
                <td key={index}>{arm}</td>
              ))
            }
          </tr>
          <tr>
            <td>Урон</td>
            {
              damage?.map((dmg, index) => (
                <td data-testid="baseDMG"
                  key={index}
                >
                  {dmg}
                </td>
              ))
            }
          </tr>
          <tr>
            <td>Критический урон</td>
            {
              criticalDamage?.map((dmg, index) => (
                <td data-testid="critDMG"
                  key={index}
                >
                  {dmg}
                </td>
              ))
            }
          </tr>
          <tr>
            <td>Урон в секунду</td>
            {
              dps?.map((dmg, index) => (
                <td data-testid="dpsDMG"
                  key={index}
                >
                  {dmg}
                </td>
              ))
            }
          </tr>
        </tbody>
      </table>
    </div>
  );
};

export default DPSTable;
