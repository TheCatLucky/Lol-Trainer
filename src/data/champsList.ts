import ChampionModel from '../models/ChampionModel';

const champsList: ChampionModel[] = [
  {
    name: 'Ahri',
    attackDamageBase: 53,
    attackDamageScale: 3,
    attackDamage: 0,
    abilityDamageBase: 0,
    abilityDamage: 0,
    attackSpeedBase: 0.668,
    attackSpeedScale: 0.02,
    attackSpeedRatio: 0.668,
    attackSpeed: 0,
    lethality: 0,
    armorPenetration: 0,
    magicFlatPenetration: 0,
    magicPenetration: 0,
    critChance: 0,
    critDamage: 1.75,
    healthBase: 500,
    healthScale: 82,
    health: 0,
    armorBase: 18,
    armorScale: 3.5,
    armor: 0,
    magicResistanceBase: 30,
    magicResistanceScale: 0.5,
    magicResistance: 0,
  },
  {
    name: 'Aatrox',
    attackDamageBase: 60,
    attackDamageScale: 5,
    attackDamage: 0,
    abilityDamageBase: 0,
    abilityDamage: 0,
    attackSpeedBase: 0.651,
    attackSpeedScale: 0.025,
    attackSpeedRatio: 0.651, // ???????????????????
    attackSpeed: 0,
    lethality: 0,
    armorPenetration: 0,
    magicFlatPenetration: 0,
    magicPenetration: 0,
    critChance: 0,
    critDamage: 1.75,
    healthBase: 580,
    healthScale: 90,
    health: 0,
    armorBase: 38,
    armorScale: 3.25,
    armor: 0,
    magicResistanceBase: 32,
    magicResistanceScale: 1.25,
    magicResistance: 0,
  },
  {
    name: 'Akali',
    attackDamageBase: 62,
    attackDamageScale: 3.3,
    attackDamage: 0,
    abilityDamageBase: 0,
    abilityDamage: 0,
    attackSpeedBase: 0.625,
    attackSpeedScale: 0.032,
    attackSpeedRatio: 0.625, // ???????????????????
    attackSpeed: 0,
    lethality: 0,
    armorPenetration: 0,
    magicFlatPenetration: 0,
    magicPenetration: 0,
    critChance: 0,
    critDamage: 1.75,
    healthBase: 500,
    healthScale: 105,
    health: 0,
    armorBase: 23,
    armorScale: 3.5,
    armor: 0,
    magicResistanceBase: 37,
    magicResistanceScale: 1.25,
    magicResistance: 0,
  },
  {
    name: 'Akshan',
    attackDamageBase: 52,
    attackDamageScale: 3.5,
    attackDamage: 0,
    abilityDamageBase: 0,
    abilityDamage: 0,
    attackSpeedBase: 0.638,
    attackSpeedScale: 0.04,
    attackSpeedRatio: 0.638, // ???????????????????
    attackSpeed: 0,
    lethality: 0,
    armorPenetration: 0,
    magicFlatPenetration: 0,
    magicPenetration: 0,
    critChance: 0,
    critDamage: 1.75,
    healthBase: 560,
    healthScale: 90,
    health: 0,
    armorBase: 26,
    armorScale: 3,
    armor: 0,
    magicResistanceBase: 30,
    magicResistanceScale: 0.5,
    magicResistance: 0,
  },
];

export default champsList;
