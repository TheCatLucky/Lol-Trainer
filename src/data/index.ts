export { default as champsList } from './champsList';

export { default as itemsList } from './itemsList';
