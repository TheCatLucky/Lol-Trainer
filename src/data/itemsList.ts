import { ItemModel } from '../models';

const itemsList: ItemModel[] = [
  {
    name: 'B.F.Sword',
    attackDamage: 55,
  },
  {
    name: 'Axiom Arc',
    attackDamage: 55,
    lethality: 10,
  },
  {
    name: 'Berserker`s Greaves',
    attackSpeed: 0.35,
  },
  {
    name: 'Black Cleaver',
    attackDamage: 45,
    health: 350,
  },
  {
    name: 'Black Mist Scythe',
    attackDamage: 20,
    health: 75,
  },
  {
    name: 'Blade of the Ruined King',
    attackDamage: 40,
    attackSpeed: 0.25,
  },
  {
    name: 'Bloodthirster ',
    attackDamage: 55,
    critChance: 0.2,
  },
  {
    name: 'Caulfield`s Warhammer ',
    attackDamage: 25,
  },
  {
    name: 'Chempunk Chainsword',
    attackDamage: 45,
    health: 250
  },
];

export default itemsList;
