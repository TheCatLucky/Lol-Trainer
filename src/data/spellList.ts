import AhriE from '../assets/Spells/Ahri_E.webp';
import AhriQ from '../assets/Spells/Ahri_Q.webp';
import AhriR from '../assets/Spells/Ahri_R.webp';
import AhriW from '../assets/Spells/Ahri_W.webp';
import { Spell } from '../models';

export const spells: {[index: string]: Spell[]} = {
  'Ahri': [
    {
      name: 'Q',
      img: AhriQ,
      baseDamage: [40, 65, 90, 115, 140],
      scaleAd: [0, 0, 0, 0, 0],
      scaleAp: [0.4, 0.4, 0.4, 0.4, 0.4],
    },
    {
      name: 'W',
      img: AhriW,
      baseDamage: [80, 120, 160, 200, 240],
      scaleAd: [0, 0, 0, 0, 0],
      scaleAp: [0.48, 0.48, 0.48, 0.48, 0.48],
    },
    {
      name: 'E',
      img: AhriE,
      baseDamage: [80, 110, 140, 170, 200],
      scaleAd: [0, 0, 0, 0, 0],
      scaleAp: [0.6, 0.6, 0.6, 0.6, 0.6],
    },
    {
      name: 'R',
      img: AhriR,
      baseDamage: [60, 90, 120],
      scaleAd: [0, 0, 0, 0, 0],
      scaleAp: [0.35, 0.35, 0.35],
    },
  ]
};