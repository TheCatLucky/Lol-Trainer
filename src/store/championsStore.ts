import { action, configure, makeObservable, observable } from 'mobx';
import { ChampionModel, SelectedItems, StatsEnum } from '../models';

configure({
  enforceActions: 'always',
});

type ChampionsToCompare = {
  champion: ChampionModel;
  id: number;
  equipment: SelectedItems;
};
//ToDo: куда-то сетить выбранного персонажа, чтобы не изменять весь стор.
class ChampionsStore {
  /**
   * Список чемпионов
   */
  champions: ChampionModel[] = [];

  champToCompare: ChampionsToCompare[] = [];

  constructor(champions: ChampionModel[]) {
    makeObservable(this, {
      champions: observable,
      champToCompare: observable,
      setChampions: action.bound,
      setChampsToCompare: action.bound,
      calcAd: action.bound,
      calcBaseAsWithLvl: action.bound,
      calcAsWithItems: action.bound,
      calcArmor: action.bound,
      calcArmFlatPen: action.bound,
      calcMagicResist: action.bound,
      calcHealth: action.bound,
      calcNewStats: action.bound,
    });
    this.setChampions(champions);
    this.setChampsToCompare(champions[0]);
  }

  /**
   * @param champions Массив чемпионов
   * @param lvl Уровень персонажа
   */
  setChampions(champions: ChampionModel[], lvl = 1): void {
    this.champions = champions.map((champion) => ({
      ...champion,
      stats: {
        ...champion.stats,
        attackDamage: this.calcAd(champion, lvl),
        attackSpeed: this.calcBaseAsWithLvl(champion, lvl),
        health: this.calcHealth(champion, lvl),
        armor: this.calcArmor(champion, lvl),
        magicResistance: this.calcMagicResist(champion, lvl),
      },
    }));
  }
  /**
   * @param champ Чемпион для сравнения
   */
  setChampsToCompare(champ: ChampionModel): void {
    this.champToCompare = [
      {
        champion: {
          ...champ
        },
        id: 1,
        equipment: {
          items: [],
          haveMythic: false,
          legendaryIDs: [],
        },
      },
      {
        champion: {
          ...champ
        },
        id: 2,
        equipment: {
          items: [],
          haveMythic: false,
          legendaryIDs: [],
        },
      },
    ];
  }
  /**
   *
   * @param championStats Статистики чемпиона
   * @param champId ID чемпиона
   * @param items Объект с  предметами
   */
  setCompareChampionsStats(
    championStats: ChampionModel,
    champId: number,
    items: SelectedItems,
  ): void {
    this.champToCompare = this.champToCompare.map((champion) => {
      if (champId === champion.id) {
        return {
          champion: {
            ...championStats
          },
          id: champId,
          equipment: {
            ...items
          },
        };
      } else {
        return {
          champion: {
            ...champion.champion
          },
          id: champion.id,
          equipment: {
            ...champion.equipment
          },
        };
      }
    });
  }
  /**
   *
   * @param baseStats Базовые статистики
   * @param items Объект с предметами
   * @param champId ID чемпиона
   * @param champLvl Уровень чемпиона
   */
  calcNewStats(
    baseStats: ChampionModel,
    items: SelectedItems,
    champId: number,
    champLvl: number,
  ): void {
    let newStats = {
      ...baseStats
    };
    items.items.forEach((item) => {
      item.stats.forEach((field) => {
        switch (field.name) {
        case StatsEnum.attackSpeed: {
          const newAS = this.calcAsWithItems(
            newStats.stats[field.name],
            field.value,
            newStats.stats.attackSpeedRatio,
          );

          return (newStats = {
            ...newStats,
            stats: {
              ...newStats.stats,
              [field.name]: newAS,
            },
          });
        }

        case StatsEnum.lethality: {
          const armorFlatPen = this.calcArmFlatPen(
            newStats.stats[field.name] + field.value,
            champLvl,
          );

          return (newStats = {
            ...newStats,
            stats: {
              ...newStats.stats,
              [field.name]: newStats.stats[field.name] + field.value,
              armorFlatPenetration: armorFlatPen,
            },
          });
        }

        default:
          return (newStats = {
            ...newStats,
            stats: {
              ...newStats.stats,
              [field.name]: newStats.stats[field.name] + field.value,
            },
          });
        }
      });
    });
    this.setCompareChampionsStats(newStats, champId, items);
  }

  /**
   * @param scale Прирост за уровень
   * @param lvl Уровень персонажа
   */
  calcGrowth(scale: number, lvl: number): number {
    return scale * (lvl - 1) * (0.7025 + 0.0175 * (lvl - 1));
  }

  /**
   * @param champion Чемпион
   * @param lvl Уровень чемпиона
   */
  calcAd(champion: ChampionModel, lvl: number): number {
    const growth = this.calcGrowth(champion.scale.attackDamage, lvl);

    return Math.round(champion.base.attackDamage + growth);
  }

  calcBaseAsWithLvl(champion: ChampionModel, lvl: number): number {
    const growth = this.calcGrowth(champion.scale.attackSpeed, lvl);

    return (
      Math.round((champion.base.attackSpeed + growth * champion.stats.attackSpeedRatio) * 1000) /
      1000
    );
  }

  //TODO переделать расчет армора с учетом новых характеристик:
  // базовый армор, текущий базовый, текущий бонусный и общий армор.
  calcArmor(champion: ChampionModel, lvl: number): number {
    const growth = this.calcGrowth(champion.scale.armor, lvl);

    return Math.round(champion.base.armor + growth);
  }

  calcMagicResist(champion: ChampionModel, lvl: number): number {
    const growth = this.calcGrowth(champion.scale.magicResistance, lvl);

    return Math.round(champion.base.magicResistance + growth);
  }

  calcHealth(champion: ChampionModel, lvl: number): number {
    const growth = this.calcGrowth(champion.scale.health, lvl);

    return Math.round(champion.base.health + growth);
  }

  calcArmFlatPen(lethality: number, lvl: number): number {
    return Math.round(lethality * (0.6 + (0.4 * lvl) / 18) * 100) / 100;
  }

  calcAsWithItems(currentAS: number, itemAS: number, asRatio: number): number {
    return Math.round((currentAS + itemAS * asRatio) * 1000) / 1000;
  }
}

export default ChampionsStore;
