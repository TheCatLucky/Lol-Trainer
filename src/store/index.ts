export { default as ChampionsStore } from './championsStore';
export { default as ItemsStore } from './itemsStore';
