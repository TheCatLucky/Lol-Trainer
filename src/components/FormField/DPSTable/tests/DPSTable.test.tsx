import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { champsList } from '../../../../data';
import { ChampionsStore } from '../../../../store';
import DPSTable from '../DPSTable';
const champsStore = new ChampionsStore(champsList);
const Ahri = (
  <BrowserRouter>
    <DPSTable champion={champsStore.champions[0]} />
  </BrowserRouter>
);
const Akshan = (
  <BrowserRouter>
    <DPSTable champion={champsStore.champions[3]} />
  </BrowserRouter>
);
describe('Компонент DPSTable Ahri', () => {
  it('отображает базовый урон', () => {
    render(Ahri);
    const damageArray = screen.getAllByTestId('baseDMG');
    expect(damageArray[0].innerHTML).toBe('53');
    expect(damageArray[10].innerHTML).toBe('26');
  });

  it('отображает критический урон', () => {
    render(Ahri);
    const damageArray = screen.getAllByTestId('critDMG');
    expect(damageArray[0].innerHTML).toBe('92');
    expect(damageArray[10].innerHTML).toBe('46');
  });
  it('отображает dps', () => {
    render(Ahri);
    const damageArray = screen.getAllByTestId('dpsDMG');
    expect(damageArray[0].innerHTML).toBe('35');
    expect(damageArray[10].innerHTML).toBe('17');
  });
});

describe('Компонент DPSTable Akshan', () => {
  it('отображает базовый урон', () => {
    render(Akshan);
    const damageArray = screen.getAllByTestId('baseDMG');
    expect(damageArray[0].innerHTML).toBe('52');
    expect(damageArray[10].innerHTML).toBe('26');
  });

  it('отображает критический урон', () => {
    render(Akshan);
    const damageArray = screen.getAllByTestId('critDMG');
    expect(damageArray[0].innerHTML).toBe('91');
    expect(damageArray[10].innerHTML).toBe('45');
  });
  it('отображает dps', () => {
    render(Akshan);
    const damageArray = screen.getAllByTestId('dpsDMG');
    expect(damageArray[0].innerHTML).toBe('33');
    expect(damageArray[10].innerHTML).toBe('16');
  });
});
