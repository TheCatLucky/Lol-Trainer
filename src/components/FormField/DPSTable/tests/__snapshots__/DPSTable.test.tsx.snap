// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Компонент DPSTable рендерится корректно 1`] = `
<BrowserRouter>
  <DPSTable
    champion={
      Object {
        "abilityDamage": 0,
        "abilityDamageBase": 0,
        "armor": 18,
        "armorBase": 18,
        "armorFlatPenetration": 0,
        "armorPenetration": 0,
        "armorScale": 3.5,
        "attackDamage": 53,
        "attackDamageBase": 53,
        "attackDamageScale": 3,
        "attackSpeed": 0.668,
        "attackSpeedBase": 0.668,
        "attackSpeedRatio": 0.668,
        "attackSpeedScale": 0.02,
        "critChance": 0,
        "critDamage": 1.75,
        "health": 500,
        "healthBase": 500,
        "healthScale": 82,
        "lethality": 0,
        "magicFlatPenetration": 0,
        "magicPenetration": 0,
        "magicResistance": 30,
        "magicResistanceBase": 30,
        "magicResistanceScale": 0.5,
        "name": "Ahri",
      }
    }
  />
</BrowserRouter>
`;
